src/main/java/de/tum/in/www1/artemis/service/connectors/athena/AthenaSubmissionSelectionService.java

```diff
@@ -0,0 +1,113 @@
+
+        RequestDTO(@NotNull TextExercise exercise, @NotNull List<Long> submissionIds) {
+            this.exercise = TextExerciseDTO.of(exercise);
+            this.submissionIds = submissionIds;
+        }
+    }
+
+    private record ResponseDTO(long data // will contain the submission ID to choose, or -1 if no submission was explicitly chosen
+    ) {
+    }
+
+    /**
+     * Create a new AthenaSubmissionSelectionService, which uses a custom timeout for requests to Athena
+     */
+    public AthenaSubmissionSelectionService(@Qualifier("athenaRestTemplate") RestTemplate athenaRestTemplate) {
+        // Configure rest template to use the given timeout
+        SimpleClientHttpRequestFactory requestFactory = new SimpleClientHttpRequestFactory();
+        requestFactory.setConnectTimeout(REQUEST_TIMEOUT_MS);
+        requestFactory.setReadTimeout(REQUEST_TIMEOUT_MS);
+        athenaRestTemplate.setRequestFactory(requestFactory);
+        // Create connector
+        connector = new AthenaConnector<>(log, athenaRestTemplate, ResponseDTO.class);
+    }
+
+    /**
+     * Fetches the proposedTextSubmission for a given exercise from Athena
+     *
+     * @param exercise    the exercise to get the proposed Submission for
+     * @param submissions assessable submissions of the exercise
+     * @return a Submission suggested by the Athena submission selector (e.g. chosen by the highest information gain)
+     * @throws IllegalArgumentException if exercise isn't automatically assessable
+     */
+    public Optional<TextSubmission> getProposedSubmission(TextExercise exercise, List<TextSubmission> submissions) {
+        if (!exercise.isFeedbackSuggestionsEnabled()) {
+            throw new IllegalArgumentException("The Exercise does not have feedback suggestions enabled.");
+        }
+        if (submissions.isEmpty()) {
+            return Optional.empty();
+        }
+
+        log.debug("Start Athena Submission Selection Service for Text Exercise '{}' (#{}).", exercise.getTitle(), exercise.getId());
+
+        log.info("Calling Remote Service to calculate next proposed submissions for {} submissions.", submissions.size());
+
+        try {
+            final RequestDTO request = new RequestDTO(exercise, submissions.stream().mapToLong(TextSubmission::getId).boxed().toList());
+            // TODO: make module selection dynamic (based on exercise)
+            // allow no retries because this should be fast and it's not too bad if it fails
+            ResponseDTO response = connector.invokeWithRetry(athenaUrl + "/modules/text/module_text_cofee/select_submission", request, 0);
+            log.info("Remote Service to calculate next proposes submissions responded: {}", response.data);
+            long submissionId = response.data;
+            if (submissionId == -1) {
+                return Optional.empty();
+            }
+            return submissions.stream().filter(s -> s.getId() == submissionId).findFirst();
```

Comment:
Please check (maybe not here but after calling the method) if the selected submission is still assessable (i.e. did not get locked in the meantime by a different user). I assume that getting the best available submission is deterministic, meaning if two tutors press "assess" at the same time, they would try to both assess the same submission.

Reply:
It will depend on the module, but checking is important regardless 👍 I've added that.